File: listnode/listnode5_lc92.py
"""
    反转链表II
    给你单链表的头指针 head 和两个整数 left 和 right ，其中 left <= right 。
    请你反转从位置 left 到位置 right 的链表节点，返回 反转后的链表 。
"""
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseBetween(self, head, left, right):
        # 两次遍历(找到left和right要遍历一次，反转要遍历一次)

        # 反转链表
        def reverse_linked_list(head):
            pre = None
            cur = head
            while cur:
                tmp = cur.next
                cur.next = pre
                pre = cur
                cur = tmp

        dummy = ListNode(0)
        dummy.next = head
        pre = dummy

        # 第一步：走到left节点的前一个节点
        for _ in range(left - 1):
            pre = pre.next
        
        # 第二步：来到right节点
        right_node = pre
        for _ in range(right - left + 1):
            right_node = right_node.next
        
        # 第三步：切割出一个子链表
        left_node = pre.next
        post = right_node.next
        pre.next = None
        right_node.next = None

        # 第四步：反转链表
        reverse_linked_list(left_node)

        # 第五步：接回到原来的链表中
        pre.next = right_node
        left_node.next = post
        return dummy.next


        # 一次遍历
        # dummy= ListNode(0)
        # dummy.next = head
        # pre = dummy
        # for i in range(left - 1):
        #     pre = pre.next
        # cur = pre.next
        # for i in range(left, right):
        #     post = cur.next
        #     cur.next = post.next
        #     post.next = pre.next      # 注意这里不是post.next = cur
        #     pre.next = post
        # return dummy.next

File: cpp/listnode/listnode5_lc92.cpp
/*
    反转链表II
    给你单链表的头指针 head 和两个整数 left 和 right ，其中 left <= right 。
    请你反转从位置 left 到位置 right 的链表节点，返回 反转后的链表 。
*/

#include <iostream>
#include <vector>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (head == NULL)
        {
            return NULL;
        }
        ListNode *temp;
        ListNode *pre = NULL;
        ListNode *cur = head;
        while (cur)
        {
            temp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = temp;
        }
        return pre;
    }

    ListNode *reverseBetween(ListNode *head, int left, int right)
    {
        ListNode *dummy = new ListNode(0, head);
        ListNode *pre = dummy;

        // 第一步：走到left节点的前一个节点
        for (int i = 0; i < left - 1; i++)
        {
            pre = pre->next;
        }

        // 第二步：来到right节点
        ListNode *right_node = pre;
        for (int i = 0; i < right - left + 1; i++)
        {
            right_node = right_node->next;
        }

        // 第三步：切割出一个子链表
        ListNode *left_node = pre->next;
        ListNode *post = right_node->next;
        pre->next = NULL;
        right_node->next = NULL;

        // 第四步：翻转链表
        reverseList(left_node);

        // 第五步：接回原来的链表中
        pre->next = right_node;
        left_node->next = post;
        return dummy->next;
    }
};

void printListNode(ListNode *head)
{
    if (head == NULL)
    {
        return;
    }
    while (head)
    {
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;
}

int main()
{
    // 构建链表
    ListNode *node1 = new ListNode(1);
    ListNode *node2 = new ListNode(2);
    ListNode *node3 = new ListNode(3);
    ListNode *node4 = new ListNode(4);
    ListNode *node5 = new ListNode(5);
    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;

    int left = 2;
    int right = 4;
    Solution *obj = new Solution();
    ListNode *reverse_head = obj->reverseBetween(node1, left, right);
    printListNode(reverse_head); // 1 4 3 2 5
    return 0;
}